wallet/cardano-trezor-crypto-provider.js

```diff
@@ -0,0 +1,370 @@
+    return await new Promise((resolve, reject) => {
+      // m/44'/1815'/0'/0/childIndex
+      const path = normalizePathsForTrezor([derivationPath])[0]
+
+      callTrezor((shouldRejectOnError) => {
+        TrezorConnect.adaGetXpubKey(path, (response) => {
+          if (response.success) {
+            const xpubData = {
+              xpub: response.xpub_key,
+              root_hd_passphrase: response.root_hd_passphrase,
+            }
+
+            if (!state.rootHdPassphrase) {
+              state.rootHdPassphrase = new Buffer(response.root_hd_passphrase, 'hex')
+            }
+
+            resolve(xpubData)
+          } else {
+            if (shouldRejectOnError(response.error)) {
+              reject(response.error)
+            }
+          }
+        })
+      })
+    })
+  }
+
+  async function deriveXpubNonHardened(derivationPath) {
+    const parentPath = derivationPath.slice(0, derivationPath.length - 1)
+    const childPath = derivationPath.slice(derivationPath.length - 1, derivationPath.length)
+
+    // this reduce ensures that this would work even for empty derivation path
+    return childPath.reduce(derivePublic, await deriveXpub(parentPath, 'hardened'))
+  }
+
+  async function isOwnAddress(address) {
+    const cachedAddress = Object.values(state.derivedAddresses).find((record) => {
+      if (record.address === address) {
+        return true
+      }
+
+      return false
+    })
+
+    if (cachedAddress) {
+      return true
+    }
+
+    try {
+      const hdPassphrase = await getRootHdPassphrase()
+      unpackAddress(address, hdPassphrase)
+      return true
+    } catch (e) {
+      if (e.name === 'AddressDecodingException') {
+        return false
+      }
+
+      throw e
+    }
+  }
+
+  function deriveHdNode(childIndex) {
+    throw new Error('This operation is not supported on TrezorCryptoProvider!')
+  }
+
+  async function sign(message, derivationPath) {
+    return await new Promise((resolve, reject) => {
+      const messageToSign = Buffer.from(message, 'hex').toString('utf8')
+
+      // m/44'/1815'/0'/0/childIndex
+      const path = normalizePathsForTrezor([derivationPath])[0]
+
+      callTrezor((shouldRejectOnError) => {
+        TrezorConnect.adaSignMessage(path, messageToSign, (response) => {
+          if (response.success) {
+            resolve(new Buffer(response.signature, 'hex'))
+          } else {
+            if (shouldRejectOnError(response.error)) {
+              reject(response.error)
+            }
+          }
+        })
+      })
+    })
+  }
+
+  async function getRootHdPassphrase() {
+    if (!state.rootHdPassphrase) {
+      const rootHdPassphrase = (await deriveTrezorXpub([])).root_hd_passphrase
+      state.rootHdPassphrase = new Buffer(rootHdPassphrase, 'hex')
+    }
+
+    return state.rootHdPassphrase
+  }
+
+  async function getDerivationPathFromAddress(address) {
+    const cachedAddress = Object.values(state.derivedAddresses).find((record) => {
```

Comment:
i would simplify this to:`Object.values(state.derivedAddresses).find((record) => record.address === address)`